doc/architecture.md

```diff
@@ -0,0 +1,136 @@
+Architecture
```

Comment:
Overall, there's a lot of good information on this page. I think we should organize it more strategically and provide signposts at the beginning that tie everything together in a cohesive way.